File: 2018/Day 8/arrows.py
with open('input.txt', 'r') as f:
    xs = [*map(int, f.read().split())]
def mk_tree(childs, metas, *xs):
    xs, ts = mk_node(childs, xs, [])
    node = (ts, xs[:metas])
    return (xs[metas:], node)
def mk_node(childs, xs, ts):
    for _ in range(childs):
        xs, node = mk_tree(*xs)
        ts.append(node)
    return (xs, ts)
tree = mk_tree(*xs)[1]
def silver(childs, metas):
    return sum(metas) + sum(silver(*ch)
        for ch in childs)
def gold(childs, metas):
    return sum(metas) if not childs else sum(
        gold(*childs[v-1]) for v in metas if v <= len(childs))
print("Silver:", silver(*tree))
print("Gold:", gold(*tree))

File: 2018/Day 08/arrows.hs
module Main where

main = do
  tree <- snd.mk_tree.map read.words <$> readFile "input.txt"
  putStrLn$ "Silver: " ++ (show $silver tree)
  putStrLn$ "Gold: " ++ (show $gold tree)

data Tree a = Node [Tree a] [a] deriving Show

mk_tree :: [Int] -> ([Int], Tree Int)
mk_tree (k:m:xs) = (drop m ys, Node (reverse ts) $take m ys)
  where (ys, ts) = foldr mk_nodes (xs, []) [1..k]

mk_nodes :: Int -> ([Int], [Tree Int]) -> ([Int], [Tree Int])
mk_nodes _ (xs,ts) = let (ys, t) = mk_tree xs in (ys, t:ts)

silver :: Tree Int -> Int
silver (Node ns vs) = sum vs + (sum $map silver ns)

gold :: Tree Int -> Int
gold (Node ns vs) = sum$ if null ns then vs else
  [gold.last$take v ns | v <- vs, v <= length ns]
